contents/testing/writing-testable-code.md

```diff
@@ -0,0 +1,110 @@
+
+## Favor composition over inheritance
+
+Because
+
+> At run-time you can not chose a different inheritance, but you can chose a different composition. This is important for tests as we want to test thing in isolation.
+
+Wrongly used inheritance hurts clutter the focus of test because you need to mock out the your parent classes' irrelevant dependencies, too.
+
+> Inheriting from AuthenticatedServlet will make your sub-class very hard to test since every test will have to mock out the authentication. And you have to do so to successfully navigate the super class. But what if AuthenticatedServlet inherits from DbTransactionServlet? (that gets so much harder)
+
+## Favor polymorphism over conditionals
+
+If a class has `switch` or the same `if` conditions in many places, you need to break down your complex class into several smaller simpler classes.
+
+Because testing smaller and simpler classes is easier, polymorphism can help you do that.
+
+## Mixing Service Objects with Value Objects
+
+Values objects are your **model** objects, like `User`, `Email`, `CreditCard`.
+
+1. They are (or should be) very easy to construct. And they should never take a service object in its constructor (since otherwise it is not easy to construct).
+2. You can freely create value objects with the "new" operator directly in line with your business logic (exception to [point 1](#Don’t-mix-object-construction-with-application-logic) since they are leafs of your application graph).
+3. They are **never mocked**.
+
+Service objects, in the other way, are your **logic** objects, like `UserAthenticator`, `MailServer`, `CreditCardProcessor`. Compared to value objects,
+
+1. Service objects are harder to construct. And their constructors ask for lots of other objects for collaboration.
+2. Service objects are harder to construct and as a result are never constructed with a new operator in-line, (instead use factory / DI-framework) for the object graph construction.
+**Note**: service objects don't take value-objects in their constructors since DI-frameworks tend to be unaware about the how to create a value-object.
+3. Service objects are harder to test since they are all about collaboration and as a result we are forced to use mocking, something which we want to minimize.
+
+Clearly, they are very different in their nature and usage.
+
+> Mixing the two creates a hybrid which has no advantages of value-objects and all the baggage of service-object.
+
+## Mixing of Concerns
+
+Single Responsibility Principle also requires you to have classes with one concern, because these classes are simpler to understand what they are doing and easier to unit test.
+
+Symptoms of mixing of concerns:
+
+* summing up what the class does includes the word "and"
+* class has fields that are only used in some methods
+* class has static methods that only operate on parameters
```

Comment:
This is a good point! Similar things like `Math.abs()`. They are **acceptable** because they are at the end of call graph. But even though they are acceptable, why write Math.abs(-5) if you can write -5.abs(). Java got this wrong. Ruby got this right. :)